### PR TITLE
IIS ANCM stress test application

### PR DESCRIPTION
**Pull request by bangbingsyb**
Add ANCM stress test app to IISIntegration so that our stress machines can directly pull the latest source from the repo, build the app locally with the latest dependencies.

Removed the dependency of the app to WebSockets middleware based on the suggestion from Chris and Justin to avoid circular dependency.

### CODE REVIEW

File: samples/ANCMStressTestSample/Startup.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        private void WebSocket(IApplicationBuilder app)
+        {
+            app.Run(async context =>
+            {
+                var upgradeFeature = context.Features.Get<IHttpUpgradeFeature>();
+
+                // Generate WebSocket response headers
+                string key = string.Join(", ", context.Request.Headers[Constants.Headers.SecWebSocketKey]);
```
Comment (Tratcher) [inline samples/ANCMStressTestSample/Startup.cs]: `string key = context.Request.Headers[Constants.Headers.SecWebSocketKey].ToString();`

File: samples/ANCMStressTestSample/Program.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                    factory.AddConsole();
+                })
+                .UseKestrel()
+                .UseStartup<Startup>()
+                .Build();
+
+            AppLifetime = (IApplicationLifetime)host.Services.GetService(typeof(IApplicationLifetime));
+            AppLifetime.ApplicationStopping.Register(
+                () => {
+                    AppLifetimeStopping = true;
```
Comment (jkotalik) [inline samples/ANCMStressTestSample/Program.cs]: Having this been a bool seems racy to me. Provide a cancellation token that will be canceled instead? Also you can pass in that cancellation token to Send/ReceiveAsync

File: samples/ANCMStressTestSample/WebSockets/Constants.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        {
+            public const string Upgrade = "Upgrade";
+            public const string UpgradeWebSocket = "websocket";
+            public const string Connection = "Connection";
+            public const string ConnectionUpgrade = "Upgrade";
+            public const string SecWebSocketKey = "Sec-WebSocket-Key";
+            public const string SecWebSocketVersion = "Sec-WebSocket-Version";
+            public const string SecWebSocketProtocol = "Sec-WebSocket-Protocol";
+            public const string SecWebSocketAccept = "Sec-WebSocket-Accept";
+            public const string SupportedVersion = "13";
```
Comment (jkotalik) [inline samples/ANCMStressTestSample/WebSockets/Constants.cs]: Why do we need SupportedVersion?

Comment (Tratcher) [reply samples/ANCMStressTestSample/WebSockets/Constants.cs]: Some of these were from request headers, the test code assumes the request headers are correct. Purge as needed.

File: samples/ANCMStressTestSample/Startup.cs (hunk 1 above)
Comment (bangbingsyb) [reply samples/ANCMStressTestSample/Startup.cs]: Good catch! Fixed.

File: samples/ANCMStressTestSample/Program.cs (hunk 2 above)
Comment (bangbingsyb) [reply samples/ANCMStressTestSample/Program.cs]: This was copied from Jeong Hwan's original test app for functional test. Yes, for stress test purpose, we need to handle application shutdown more carefully. Made a change to how we handle close handshake.

File: samples/ANCMStressTestSample/WebSockets/Constants.cs (hunk 3 above)
Comment (bangbingsyb) [reply samples/ANCMStressTestSample/WebSockets/Constants.cs]: Purged the unused constants.

File: samples/ANCMStressTestSample/WebSockets/Constants.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+namespace ANCMStressTestApp
+{
+    public static class Constants
+    {
+        public static class Headers
+        {
+            public const string Upgrade = "Upgrade";
+            public const string UpgradeWebSocket = "websocket";
+            public const string Connection = "Connection";
+            public const string ConnectionUpgrade = "Upgrade";
```
Comment (pan-wang) [inline samples/ANCMStressTestSample/WebSockets/Constants.cs]: the value is same Upgrade. You can remove this one

File: samples/ANCMStressTestSample/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                var buffer = new byte[1024 * 4];
+                var result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), token);
+                bool closeFromServer = false;
+                string closeFromServerCmd = "CloseFromServer";
+                int closeFromServerLength = closeFromServerCmd.Length;
+
+                while (!result.CloseStatus.HasValue && !token.IsCancellationRequested && !closeFromServer)
+                {
+                    if (result.Count == closeFromServerLength &&
+                        System.Text.Encoding.ASCII.GetString(buffer).Substring(0, result.Count) == closeFromServerCmd)
```
Comment (pan-wang) [inline samples/ANCMStressTestSample/Startup.cs]: do you  want it be case sensitive?

Comment (bangbingsyb) [reply samples/ANCMStressTestSample/Startup.cs]: Yes, it is better to be case sensitive. For a regular data frame, the stress client will generate random data with random length within the given length range. So there is a very small chance that the generated data frame can be identical to the special "server close request" frame. Making it case sensitive will further reduce such chance.

File: samples/ANCMStressTestSample/WebSockets/Constants.cs (hunk 4 above)
Comment (bangbingsyb) [reply samples/ANCMStressTestSample/WebSockets/Constants.cs]: Removed

File: samples/ANCMStressTestSample/Program.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+{
+    public class Program
+    {
+        public static IApplicationLifetime AppLifetime;
+        public static CancellationTokenSource Cts = new CancellationTokenSource();
+
+        public static void Main(string[] args)
+        {
+            var host = new WebHostBuilder()
+                .ConfigureLogging((_, factory) =>
```
Comment (jkotalik) [inline samples/ANCMStressTestSample/Program.cs]: Don't you need to call UseIISIntegration()?

File: samples/ANCMStressTestSample/Startup.cs (hunk 5 above)
Comment (jkotalik) [reply samples/ANCMStressTestSample/Startup.cs]: using System.Text; and then condense this to Encoding.ASCII.GetString();

Comment (bangbingsyb) [reply samples/ANCMStressTestSample/Startup.cs]: Changed.

File: samples/ANCMStressTestSample/Program.cs (hunk 6 above)
Comment (bangbingsyb) [reply samples/ANCMStressTestSample/Program.cs]: I think we need it. Looks like the IISSample doesn't call it, either. What's the reason?

File: test/ANCMStressTestApp/Startup.cs [hunk 7]
```diff
@@ -171,7 +171,9 @@ private void WebSocket(IApplicationBuilder app)
                 // Get the WebSocket object
                 var ws = WebSocketProtocol.CreateFromStream(opaqueTransport, isServer: true, subProtocol: null, keepAliveInterval: TimeSpan.FromMinutes(2));
 
-                await Echo(ws, Program.Cts.Token);
+                var appLifetime = app.ApplicationServices.GetRequiredService<IApplicationLifetime>();
+
+                await Echo(ws, appLifetime.ApplicationStopping);
```
Comment (bangbingsyb) [inline test/ANCMStressTestApp/Startup.cs]: Looks good. So the IApplicationLifetime.ApplicationStopping itself is a cancellation token.

**Comment by pakrym** [comment]
We should add correct launchSettings and `  <Import Project="..\..\build\testsite.props" />` to every test app so it is F5 runable.

**Pull request merged**